File: Algorithms/5. Longest Palindromic Substring/swift/solution.manacher.swift
class Solution {
    func longestPalindrome(_ s: String) -> String {
        let expandStr = "^#" + s.map { String($0) }.joined(separator: "#") + "#$"
        let n = expandStr.count
        var P = [Int](repeating: 0, count: n)
        var center = 0
        var right = 0
        for i in 1..<n - 1 {
            let iMirror = 2 * center - i
            if right > i {
                P[i] = min(right - i, P[iMirror])
            }
            while expandStr[expandStr.index(expandStr.startIndex, offsetBy: i + 1 + P[i])]
                == expandStr[expandStr.index(expandStr.startIndex, offsetBy: i - 1 - P[i])]
            {
                P[i] += 1
            }
            if i + P[i] > right {
                center = i
                right = i + P[i]
            }
        }
        let maxLen = P.max() ?? 0
        let centerIndex = P.firstIndex(of: maxLen) ?? 0
        let start = s.index(s.startIndex, offsetBy: (centerIndex - maxLen) / 2)
        let end = s.index(s.startIndex, offsetBy: (centerIndex + maxLen) / 2)
        return String(s[start..<end])
    }
}

File: Algorithms/5. Longest Palindromic Substring/kotlin/solution.manacher.kt
class Solution {
    fun longestPalindrome(s: String): String {
        val expandStr = "^#" + s.toCharArray().joinToString("#") + "#$"
        val n = expandStr.length
        val P = IntArray(n)
        var center = 0
        var right = 0
        for (i in 1 until n - 1) {
            val iMirror = 2 * center - i
            if (right > i) {
                P[i] = Math.min(right - i, P[iMirror])
            }
            while (expandStr[i + 1 + P[i]] == expandStr[i - 1 - P[i]]) {
                P[i]++
            }
            if (i + P[i] > right) {
                center = i
                right = i + P[i]
            }
        }
        val maxLen = P.max() ?: 0
        val centerIndex = P.indexOf(maxLen)
        return s.substring((centerIndex - maxLen) / 2, (centerIndex + maxLen) / 2)
    }
}
